hoomd/data/data_structures.py

```diff
@@ -0,0 +1,497 @@
+from abc import ABCMeta, abstractmethod
+from collections.abc import MutableMapping, MutableSequence, MutableSet
+from contextlib import contextmanager
+from copy import deepcopy
+from itertools import cycle
+
+from hoomd.data.typeconverter import (
+    TypeConversionError, TypeConverterValue, OnlyIf, Either,
+    TypeConverterMapping, TypeConverterSequence)
+
+
+class _HOOMDDataStructures(metaclass=ABCMeta):
+    """Base class for HOOMD's base container classes.
+
+    These child classes exist to enable parameter validation and mantain
+    consistency with C++ after attaching.
+    """
+    @abstractmethod
+    def to_base(self):
+        """Cast the object into the corresponding native Python container type.
+        """
+        pass
+
+    @contextmanager
+    def _buffer(self):
+        """Allows buffering during modifications to prevent needless updates.
+
+        This is a simple system to prevent unnecessary C++ - Python
+        communication.
+        """
+        self._buffered = True
+        yield None
+        self._buffered = False
+
+    def _update(self):
+        """Signal that the object has been updated."""
+        if self._buffered:
+            return
+        elif parent is not None:
+            self._parent._handle_update(self, self._label)
+
+
+def _get_inner_typeconverter(type_def, desired_type):
+    """Handles more complicated data structure specification.
+
+    When using nested containers where an entry could be a list or None for
+    instance requires the use of something like typeconverter's OnlyIf or
+    Either. We autdetect the internal type specification for these object and if
+    they are of the correct type use these.
```

Comment:
I'm having trouble parsing what is meant here. I tried writing a suggestion but I don't know how to resolve the antecedents of "them" / "these" / etc.
```suggestion
    When using nested containers, where an entry could be a list or None for
    instance requires the use of something like typeconverter's OnlyIf or
    Either. We detect the internal type specification for these objects and if
    they are of the correct type, use these.
```